### ISSUE TITLE
Better logging

### ISSUE THREAD

**Issue by jasoncouture** [rust-osdev/bootloader#306]
The UEFI and bios loggers use only the framebuffer.

This can be an issue during headless runs (server with no display, unit tests)

I propose we refactor the logger to support multiple methods of logging output. (Serial being the first idea), with support in the kernel configuration to enable/disable sources. Or a text config file (this is likely better, so it can be configured at runtime)

This would also allow us to separate the logging and framebuffer init, as well be able to add/remove loggers at will. Rather than using `writeln` for early logging, we will be able to initialize the logger before anything else, and configure it with the optional `config.txt` or a default config that matches the current behavior.

**Comment by phil-opp** [comment]
Good idea! I fully agree that optional serial logging would be useful.

> Or a text config file (this is likely better, so it can be configured at runtime)

You mean a separate text file that is placed on the FAT partition? I guess we could create such a file in the `create_disk_image` function. The question is how we split the options into kernel and runtime options. Candidates for runtime options could be the recently merged #303 and also the framebuffer size requirements.

**Comment by phil-opp** [comment]
Sounds reasonable to me! How about we start with a kernel config option for configuring the logger and then do the runtime config file separately?

**Comment by asensio-project** [comment]
I think that the most of the crates that can be helpful for interpreting a configuration file depend on alloc. And if we create our own crate, we could want allocation.

So, what about implementing first the allocation in the bootloader.

**Comment by asensio-project** [comment]
I am working in this and I think that we should move the code of the framebuffer to ``common/src/framebuffer.rs`` and the code related to the logger to ``common/src/logger.rs``.

This will be alright for then adding the serial being and modifying the code of the logger for including this method.

Do you like this aproach or do you have another idea?

**Comment by phil-opp** [comment]
Thanks for tackling this! 

> I think that we should move the code of the framebuffer to `common/src/framebuffer.rs` and the code related to the logger to `common/src/logger.rs`.
> 
> This will be alright for then adding the serial being and modifying the code of the logger for including this method.

Sounds good!
